include/exadg/fluid_structure_interaction_precice/precice_adapter.h

```diff
@@ -0,0 +1,371 @@
+   *             variables are passed into the vector must be the same for
+   *             both functions.
+   */
+  void
+  save_current_state_if_required(const std::function<void()> & save_state);
+
+  /**
+   * @brief      Reloads the previously stored variables in case of an implicit
+   *             coupling. The current implementation supports subcycling,
+   *             i.e. previously refers o the last time
+   *             @p save_current_state_if_required() has been called.
+   *
+   * @param[out] state_variables Vector containing all variables to reload
+   *             as reference
+   *
+   * @note       This function only makes sense, if the state variables have been
+   *             stored by calling @p save_current_state_if_required. Therefore,
+   *             the order, in which the variables are passed into the
+   *             vector must be the same for both functions.
+   */
+  void
+  reload_old_state_if_required(const std::function<void()> & reload_old_state);
+
+
+  void
+  write_data(const std::string & write_mesh_name,
+             const std::string & write_data_name,
+             const VectorType &  write_data,
+             const double        computed_timestep_length);
+
+  std::vector<Tensor<1, dim>>
+  read_block_data(const std::string & mesh_name, const std::string & data_name) const;
+
+  /**
+   * @brief is_coupling_ongoing Calls the preCICE API function isCouplingOnGoing
+   *
+   * @return returns true if the coupling has not yet been finished
+   */
+  bool
+  is_coupling_ongoing() const;
+
+  /**
+   * @brief is_time_window_complete Calls the preCICE API function isTimeWindowComplete
+   *
+   * @return returns true if the coupling time window has been completed in the current
+   *         iteration
+   */
+  bool
+  is_time_window_complete() const;
+
+
+
+private:
+  // public precice solverinterface, needed in order to steer the time loop
+  // inside the solver.
+  std::shared_ptr<precice::SolverInterface> precice;
+  /// The objects handling reading and writing data
+  std::map<std::string, std::shared_ptr<CouplingInterface<dim, data_dim, VectorizedArrayType>>>
+    writer;
+  std::map<std::string, std::shared_ptr<CouplingInterface<dim, data_dim, VectorizedArrayType>>>
+    reader;
+
+  // Container to store time dependent data in case of an implicit coupling
+  std::vector<VectorType> old_state_data;
+};
+
+
+
+template<int dim, int data_dim, typename VectorType, typename VectorizedArrayType>
+template<typename ParameterClass>
+Adapter<dim, data_dim, VectorType, VectorizedArrayType>::Adapter(const ParameterClass & parameters)
+
+{
+  precice =
+    std::make_shared<precice::SolverInterface>(parameters.participant_name,
+                                               parameters.config_file,
+                                               Utilities::MPI::this_mpi_process(MPI_COMM_WORLD),
+                                               Utilities::MPI::n_mpi_processes(MPI_COMM_WORLD));
```

Comment:
why MPI_COMM_WORLD?

Reply:
Why need to initialize preCICE using a preCICE communicator. Do you prefer something different?

Reply:
I think I made this comment because we do not want to use MPI_COMM_WORLD hardcoded in ExaDG, but pass the communicator to the classes/functions if they need a communicator